pkg/spdk/engine.go

```
@@ -1229,6 +1309,379 @@ func (e *Engine) validateAndUpdateReplicaNvme(replicaName string, bdev *spdktype
+	}
+
+	defer func() {
+		if e.isExpanding {
+			e.lastExpansionFailedAt = ""
+			e.lastExpansionError = ""
+		} else if err != nil {
+			e.lastExpansionFailedAt = time.Now().UTC().Format(time.RFC3339Nano)
+			e.lastExpansionError = errors.Wrap(err, "engine failed to start expansion").Error()
+		}
+	}()
+
+	e.isExpanding = true
+	return nil
+}
+
+func (e *Engine) finishExpansion(expanded bool, size uint64) {
+	if expanded {
+		if e.lastExpansionError != "" {
+			logrus.Infof("Succeeded to expand from size %v to %v but there are some replica expansion failures: %v", e.SpecSize, size, e.lastExpansionError)
+		} else {
+			logrus.Infof("Succeeded to expand from size %v to %v", e.SpecSize, size)
+		}
+		e.SpecSize = size
+	} else {
+		logrus.Infof("Failed to expand from size %v to %v", e.SpecSize, size)
+	}
+	e.isExpanding = false
+}
+
+func (e *Engine) isAnyReplicaUnderRebuilding(replicaClients map[string]*client.SPDKClient) bool {
+	for replicaName, replicaClient := range replicaClients {
+		replica, err := replicaClient.ReplicaGet(replicaName)
+		if err != nil {
+			e.log.Warnf("Replica get failed %v", err)
+			continue
+		}
+
+		if replica.Rebuilding {
+			e.log.Warnf("Replica %s is rebuilding, abort expansion", replicaName)
+			return true
+		}
+	}
+	return false
+}
+
+func (e *Engine) prepareRaidForExpansion(spdkClient *spdkclient.Client) (suspendFrontend bool, bdevUUID string, err error) {
+	// check if bdev raid exist
+	bdevRaid, err := spdkClient.BdevRaidGet(e.Name, 0)
+	if err != nil {
+		if jsonrpc.IsJSONRPCRespErrorNoSuchDevice(err) {
+			return false, "", nil // RAID does not exist, nothing to do
+		}
+		return false, "", errors.Wrap(err, "failed to get bdev raid")
+	}
+	if len(bdevRaid) == 0 {
+		return false, "", nil // RAID already deleted
+	}
+	bdevUUID = bdevRaid[0].UUID
+
+	// Suspend IO if frontend is active
+	if e.Frontend == types.FrontendSPDKTCPBlockdev && e.Endpoint != "" {
```

For now, need to return an error when the frontend is ublk